ui/server/src/main/scala/pl/touk/nussknacker/ui/validation/ProcessValidation.scala

```diff
@@ -9,61 +9,71 @@ import pl.touk.nussknacker.engine.api.expression.ExpressionParser
 import pl.touk.nussknacker.engine.canonicalgraph.CanonicalProcess
 import pl.touk.nussknacker.engine.compile.{NodeTypingInfo, ProcessValidator}
 import pl.touk.nussknacker.engine.graph.node.{Disableable, NodeData, Source, SubprocessInputDefinition}
+import pl.touk.nussknacker.engine.util.cache.{CacheConfig, DefaultCache}
 import pl.touk.nussknacker.restmodel.displayedgraph.DisplayableProcess
 import pl.touk.nussknacker.restmodel.displayedgraph.displayablenode.Edge
 import pl.touk.nussknacker.restmodel.process.ProcessingType
 import pl.touk.nussknacker.restmodel.validation.ValidationResults.{NodeTypingData, ValidationResult}
 import pl.touk.nussknacker.restmodel.validation.{CustomProcessValidator, PrettyValidationErrors}
 import pl.touk.nussknacker.ui.definition.UIProcessObjectsFactory
+import pl.touk.nussknacker.ui.process.ProcessCategoryService.Category
 import pl.touk.nussknacker.ui.process.marshall.ProcessConverter
 import pl.touk.nussknacker.ui.process.processingtypedata.ProcessingTypeDataProvider
 import pl.touk.nussknacker.ui.process.subprocess.SubprocessResolver
-import shapeless.syntax.typeable._
 
 object ProcessValidation {
-  def apply(data: ProcessingTypeDataProvider[ModelData],
+  def apply(modelData: ProcessingTypeDataProvider[ModelData],
             additionalProperties: ProcessingTypeDataProvider[Map[String, AdditionalPropertyConfig]],
             subprocessResolver: SubprocessResolver,
             customProcessNodesValidators: ProcessingTypeDataProvider[CustomProcessValidator]): ProcessValidation = {
-    new ProcessValidation(data.mapValues(_.validator), additionalProperties, subprocessResolver, customProcessNodesValidators)
+    new ProcessValidation(modelData, additionalProperties, subprocessResolver, customProcessNodesValidators, None)
   }
 }
 
-class ProcessValidation(validators: ProcessingTypeDataProvider[ProcessValidator],
+class ProcessValidation(modelData: ProcessingTypeDataProvider[ModelData],
                         additionalPropertiesConfig: ProcessingTypeDataProvider[Map[String, AdditionalPropertyConfig]],
                         subprocessResolver: SubprocessResolver,
-                        customProcessNodesValidators: ProcessingTypeDataProvider[CustomProcessValidator]) {
+                        customProcessNodesValidators: ProcessingTypeDataProvider[CustomProcessValidator],
+                        expressionParsers: Option[PartialFunction[ExpressionParser, ExpressionParser]]) {
 
   val uiValidationError = "UiValidation"
 
+  private val processValidatorCache = new DefaultCache[String, ProcessValidator](CacheConfig())
```

Comment:
this is more tricky than it seems... ProcessingTypeDataProvider can handle model data reload. Can we try to do it as `validators: ProcessingTypeDataProvider[DefaultCache[String, ProcessValidator]]`?